app/service/contact_svc.py

```diff
@@ -151,3 +152,17 @@ def __init__(self, agent_config):
             cmd = self.encode_string(agent.replace(i.test))
             instructions.append(Instruction(identifier=new_id, command=cmd, executor=i.executor))
         return instructions
+
+    async def _add_agent_to_operation(self, agent):
+        """Determine which operation(s) incoming agent belongs to and
+        add it to operation.
+
+        Note: Agent is added immediately to operation, as certain planners
+        may execute single links at a time before relinquishing control back
+        to c_operation.run() (when previously the operation was updated with
+        new agents), and during those link executions, new agents may arise
+        which the planner needs to be aware of.
+        """
+        for op in await self.get_service('data_svc').locate('operations', match=dict(finish=None)):
+            if op.group == agent.group or op.group is None:
+                op.update_operation(self.get_services())
```

Comment:
we should avoid adding the agent if it already exists on the operation

Reply:
Dont these linses prevent that:

 https://github.com/mitre/caldera/blob/master/app/service/contact_svc.py#L89-L94

Reply:
i think you're right-- those lines don't really handle it-- but where you put your _add_agent_to_operation call will only get hit on an agent's first beacon, so i think you're covering it the right way. disregard